### ISSUE TITLE
Do a holistic renaming of the various predicates

### ISSUE THREAD

**Issue by TWal** [REPROSEC/dolev-yao-star-extrinsic#22]
As noticed by @cwaldm when discussing on #19, we have many predicate types and it would be nice to have some consistency between them.

I think in general the types are named `foo_predicate`, and the predicate is named either:
- `bar_pred`, `baz_pred` when there are several sub-predicates
- `pred` when there is only one predicate

There is also the question of `predicate` vs `invariant`. I think I tried to use `predicate` for types that actually contain the predicates, and `invariant` for things that are more collection of predicates.

In `Core`:
https://github.com/REPROSEC/dolev-yao-star-extrinsic/blob/452210f88ac1180710f93ebecbb6394e38b970fa/src/core/DY.Core.Bytes.fst#L237-L267
https://github.com/REPROSEC/dolev-yao-star-extrinsic/blob/452210f88ac1180710f93ebecbb6394e38b970fa/src/core/DY.Core.Trace.Invariant.fst#L30-L50
https://github.com/REPROSEC/dolev-yao-star-extrinsic/blob/452210f88ac1180710f93ebecbb6394e38b970fa/src/core/DY.Core.Trace.Invariant.fst#L54-L58
https://github.com/REPROSEC/dolev-yao-star-extrinsic/blob/452210f88ac1180710f93ebecbb6394e38b970fa/src/core/DY.Core.Trace.Invariant.fst#L66-L70

In `DY.Lib.Event`:
https://github.com/REPROSEC/dolev-yao-star-extrinsic/blob/452210f88ac1180710f93ebecbb6394e38b970fa/src/lib/event/DY.Lib.Event.Typed.fst#L41-L42

In `DY.Lib.State`:
https://github.com/REPROSEC/dolev-yao-star-extrinsic/blob/452210f88ac1180710f93ebecbb6394e38b970fa/src/lib/state/DY.Lib.State.Tagged.fst#L25-L41
https://github.com/REPROSEC/dolev-yao-star-extrinsic/blob/452210f88ac1180710f93ebecbb6394e38b970fa/src/lib/state/DY.Lib.State.Typed.fst#L22-L38
https://github.com/REPROSEC/dolev-yao-star-extrinsic/blob/452210f88ac1180710f93ebecbb6394e38b970fa/src/lib/state/DY.Lib.State.Map.fst#L27-L37

In particular, @cwaldm proposed to rename `state_predicate` into `global_state_predicate` to oppose it to `local_state_predicate`. I would rather keep the names of DY.Core agnostic to what is done in DY.Lib and keep the name as-is, but add some documentation above `state_predicate` to explain that it is a global predicate and that separating it into several local state predicates is done in DY.Lib.

**Comment by qaphla** [comment]
I would argue that our crypto predicates (e.g., `aead_pred`) are not invariants, at least not in the same sense as the other ones, in that they don't say anything at the trace level. I suppose they're a bytes-level invariant of sorts, but even then, I think the actual invariant is something derived from `aead_pred`, rather than `aead_pred` itself --- in particular, I guess I would say that the corresponding invariant to `aead_pred` is the `AeadEnc` case of `bytes_invariant`:

https://github.com/REPROSEC/dolev-yao-star-extrinsic/blob/452210f88ac1180710f93ebecbb6394e38b970fa/src/core/DY.Core.Bytes.fst#L328-L355

Specifically, lines 339-355.

Packaging an invariant with its `later` lemma makes sense to me, though I'm not sure how we'd avoid writing the `later` lemma in doing so, unless you mean avoiding needing to call it so much (which is also good).

**Comment by TWal** [comment]
@qaphla Good catch. Indeed, it looks like we could renaming everything "invariant" except the "crypto predicates".

@cwaldm To use a generic "predicate that is preserved by trace extension" type, we would need to uncurry all the predicates, e.g. the state invariant would become `trace -> (principal & nat & bytes) -> prop` instead of the current `trace -> principal -> nat -> bytes -> prop`. I don't think it's worth trying to factorize this too much, I think the actual state of things is readable enough, and it would only benefit the crypto predicates, because all the state invariants also have the `pred_knowable` constraint that is difficult to generalize.

**Comment by TWal** [comment]
I am looking to clean old issues, and it seems that the result of the discussion on this one indicates things are good as-is and we can close without further action?

**Comment by cwaldm** [comment]
I think, I still prefer to explicitly call the invariants invariants. So that it gets clear from the name that the predicates inside have the invariant property. (`predicate` is pretty general)


To me, it doesn't make much sense to call a collection of predicates invariants.

I get the point, that the crypto predicates are not (or at least) different invariants than the other.

So, I'd prefer to use `invariant` for those predicates that are actual trace invariants, and leave the others as predicates.
 I.e., `state_predicate` should be `state_invariant`, but I'm fine with leaving the property as `pred` (but propose `pred_is_invariant` or something like this). Same for `local_state_predicate` and `map_predicate`.
This would make much clearer for example, that the `event_predicate` does NOT need to be an invariant, in contrast to the `state_predicate`.
Which raises the question whether `trace_invariants` is a good name, as it contains the non-invariant `event_pred`?

Or we use `predicate` everywhere. Though, I'm not sure what that means for `bytes_invariant` for example.

All in all this is not blocking as everyone else seems to be fine with the current state of things. I just think, it would make DY* a bit easier to read and understand since we have the different predicates of which only some are invariants.
